### DISCUSSION TITLE
No GPU detected on M2 Mac

### DISCUSSION THREAD

**Discussion by siddhsql** [deepjavalibrary/djl#3160] [Q&A]
Hello,

I am trying to use this library. I have added following dependencies:

```
<!-- https://mvnrepository.com/artifact/ai.djl/api -->
<dependency>
    <groupId>ai.djl</groupId>
    <artifactId>api</artifactId>
    <version>0.27.0</version>
</dependency>
<!-- https://mvnrepository.com/artifact/ai.djl.huggingface/tokenizers -->
<dependency>
    <groupId>ai.djl.huggingface</groupId>
    <artifactId>tokenizers</artifactId>
    <version>0.27.0</version>
</dependency>
<!-- https://mvnrepository.com/artifact/ai.djl.pytorch/pytorch-engine -->
<dependency>
    <groupId>ai.djl.pytorch</groupId>
    <artifactId>pytorch-engine</artifactId>
    <version>0.27.0</version>
</dependency>    
```

and [this](https://d2l.djl.ai/chapter_deep-learning-computation/use-gpu.html) code:

```
public static Device[] tryAllGpus() {
    int gpuCount = Engine.getInstance().getGpuCount();
    System.out.println("GPU count = " + gpuCount);
    if (gpuCount > 0) {
        Device[] devices = new Device[gpuCount];
        for (int i = 0; i < gpuCount; i++) {
            devices[i] = Device.gpu(i);
        }
        return devices;
    }
    return new Device[]{Device.cpu()};
}
```

which gives me

```
GPU count = 0
```

on a Mac with M2 chip. so my question is how can I leverage GPU with this library on a Mac?

**Comment by frankfliu** [comment]
The GPU device on M2 is called "mps", you have to use:

```
        Device device = Device.of("mps", -1);
```

Currently it has many limitations with Torch. We cannot make it default. see: https://github.com/deepjavalibrary/djl/pull/2044

**Comment by siddhsql** [comment]
Thanks Frank. I tried it (mps) to calculate embeddings using sentence
transformers all_mini_lm and the performance was actually a bit worse than
using cpu. could this be because of the limitations you are talking about
above? i did not get any exceptions. just slower performance.
